File: js/test/Exchange/test.watchTrades.js
'use strict'

// ----------------------------------------------------------------------------

const log = require ('ololog')
    , chai = require ('chai')
    , asTable = require ('as-table')
    , assert = chai.assert
    , testTrade = require ('ccxt/js/test/Exchange/test.trade.js')
    , errors = require ('ccxt/js/base/errors.js')

/*  ------------------------------------------------------------------------ */

module.exports = async (exchange, symbol) => {

    // log (symbol.green, 'watching trades...')

    const method = 'watchTrades'

    // we have to skip some exchanges here due to the frequency of trading
    const skippedExchanges = [
        'binanceje',
        'bitvavo',
        'currencycom',
        'dsx',
        'idex2', // rinkeby testnet, trades too rare
        'ripio',
    ]

    if (skippedExchanges.includes (exchange.id)) {
        log (exchange.id, method, 'test skipped')
        return
    }

    if (!exchange.has[method]) {
        log (exchange.id, 'does not support', method + '() method')
        return
    }

    let response = undefined

    let now = Date.now ()
    const ends = now + 15000

    while (now < ends) {

        try {

            response = await exchange[method] (symbol)

            now = Date.now ()

            assert (response instanceof Array)

            log (exchange.iso8601 (now), exchange.id, symbol.green, method, Object.values (response).length.toString ().green, 'trades')

            // log.noLocate (asTable (response))

            for (let i = 0; i < response.length; i++) {
                testTrade (exchange, response[i], symbol, now)
            }
        } catch (e) {

            if (!(e instanceof errors.NetworkError)) {
                throw e
            }

            now = Date.now ()
        }
    }

    return response
}

File: python/test/exchange/test_watch_trades.py
# -*- coding: utf-8 -*-

from exchange.test_trade import test_trade
# from exchange.table import table


async def test_watch_trades(exchange, symbol):
    # todo add real ohlcv tests here
    # log (symbol.green, 'watching ohlcvs...')
    method = 'watchTrades'
    if (method in exchange.has) and exchange.has[method]:
        response = None
        now = exchange.milliseconds()
        end = now + 20000
        while now < end:
            # print('-----------------------------------------------------------')
            trades = await getattr(exchange, method)(symbol)
            now = exchange.milliseconds()
            print(exchange.iso8601(now), symbol, len(trades), 'trades')
            for trade in trades:
                test_trade(exchange, trade, method, symbol)
            # print(table([exchange.omit(t, ['info', 'timestamp']) for t in trades]))
        return response
    else:
        print(exchange.id, method, 'is not supported or not implemented yet')


__all__ = ['test_watch_trades']
